feat (algo): Add MAP-Elites Low-Spread

btjanaka opened the pull request:
Related issues: [refer to issues] Resolves #151 

[Introduce the overall change made in the PR and list the associated modifications below]

Add MAP-Elites Low-Spread: https://dl.acm.org/doi/abs/10.1145/3583131.3590433

This PR introduces:
- `MELSRepertoire` in `mels_repertoire.py`
  - Tests in `mels_repertoire_test.py`
- `MELS` in `mels.py`
  - Tests in `mels_test.py`
- `Spread` type in `types.py`
- `multi_sample_scoring_function` in `sampling.py`

## Questions

1. Should we store all `n_evals` fitnesses and descriptors in the repertoire, or should we just store the aggregated fitness and descriptors? We can also store both the aggregate and the separate values. Storing separate values would require passing in `n_evals` to `MELSRepertoire.init_default` so that we can define the shapes of `fitnesses` and `descriptors`.
2. Should we check if one individual dominates the others if inserting multiple individuals into the same cell?
3. What link should I use for citations? Currently I am using the ACM Digital Library link. I can also use the arXiv link: https://arxiv.org/abs/2303.16207 if preferred.

## Checks

- [x] a clear description of the PR has been added
- [x] sufficient tests have been written
- [x] relevant section added to the documentation
- [x] example notebook added to the repo
- [x] clean docstrings and comments have been written
- [x] if any issue/observation has been discovered, a new issue has been opened

## Future improvements

[List here potential observations made and/or improvements that could be made in the future. If relevant, open issues for those.]

Lookatator commented:
Hi @btjanaka ! Thank you very much for this (super well-documented) PR.

I added @Egiob and @manon-but-yes as reviewers for this PR; they are probably the best people to judge it.

I had an overview of the code and it looks great! I have mainly one high-level comment regarding the implementation of the reevaluation scoring function (and where it should be implemented). I will add it as a comment.

btjanaka commented on qdax/core/containers/mels_repertoire.py:
```diff
@@ -0,0 +1,334 @@
+    # See here for how these computations work:
+    # https://jaykmody.com/blog/distance-matrices-with-numpy/
+
+    # This has the same effect as taking the dot product of each row with
+    # itself.
+    x2 = jnp.sum(descriptors**2, axis=1)
+    y2 = x2
+
+    # We can compute all x_i * y_j and store it in a matrix at xy[i][j] by
+    # taking the matrix multiplication between X and X_train transpose
+    # if you're struggling to understand this, draw out the matrices and
+    # do the matrix multiplication by hand.
+    # (m, d) x (d, n) -> (m, n)
+    xy = descriptors @ descriptors.T
+
+    # each row in xy needs to be added with x2[i]
+    # each column of xy needs to be added with y2[j]
+    # to get everything to play well, we'll need to reshape
+    # x2 from (m) -> (m, 1), numpy will handle the rest of the broadcasting for us
+    # see: https://numpy.org/doc/stable/user/basics.broadcasting.html
+    x2 = x2.reshape(-1, 1)
+    dists = jnp.sqrt(
+        x2 - 2 * xy + y2
+    )  # (m, 1) repeat columnwise + (m, n) + (n) repeat rowwise -> (m, n)
```
Good point; I believe the one-liner is `jnp.linalg.norm(x[:, None] - y)`. I'll update accordingly.

btjanaka replied:
*actually, it's `jnp.linalg.norm(x[:, None] - y, axis=2)`

btjanaka commented:
Hi @Lookatator, thank you for taking a look at this PR! I addressed your comment on the pairwise distance calculation, but I do not see the comment on scoring function evaluation. Did you make the comment in the tests or the notebook? (the scoring function is in both those places).

Egiob commented:
Hi guys! I will do a pass by the end of the week :smile:

Lookatator commented on tests/core_test/mels_test.py:
```diff
@@ -0,0 +1,196 @@
+    # Prepare the scoring function
+    bd_extraction_fn = environments.behavior_descriptor_extractor[env_name]
+
+    def scoring_fn(
+        policies_params: Genotype,
+        random_key: RNGKey,
+    ) -> Tuple[Fitness, Descriptor, ExtraScores, RNGKey]:
+        # Repeat each policy so that it can be evaluated num_evals times.
+        policies_params = jax.tree_util.tree_map(
+            lambda x: jnp.repeat(x, num_evals, axis=0), policies_params
+        )
+
+        # Call the Brax scoring function, taking into account previously defined
+        # inputs like `init_states`.
+        fitnesses, descriptors, extra_scores, random_key = scoring_function_brax_envs(
+            policies_params=policies_params,
+            random_key=random_key,
+            init_states=init_states,
+            episode_length=episode_length,
+            play_step_fn=play_step_fn,
+            behavior_descriptor_extractor=bd_extraction_fn,
+        )
+
+        # Reshape results for ME-LS.
+        extra_scores["transitions"] = jax.tree_util.tree_map(
+            lambda x: x.reshape((batch_size, num_evals, *x.shape[1:])),
+            extra_scores["transitions"],
+        )
+        return (
+            fitnesses.reshape((batch_size, num_evals)),
+            descriptors.reshape((batch_size, num_evals, -1)),
+            extra_scores,
+            random_key,
+        )
```
I discussed the way the scoring function is defined here with @limbryan,  and we think that `scoring_fn` should be as general as possible. For example, we should define a `scoring_fn` that could be passed to `MELS` but also to the standard `MAPElites` class.

To do that, I suggest that you:
- pass the standard `scoring_fn` to the `MELS` class.
- and in the `MELS` class, you override `update(...)` with a code that does a `vmap` on the scoring fn to evaluate it on several individuals. Here is an approximate suggestion on how to change `scoring_function` call in the `update(...)` method.
```python
...
random_key, *subkeys = jax.random.split(random_key, self.num_evals + 1)  # num_evals can be an attribute of the `MELS` class
fitnesses, descriptors, extra_scores, _ = jax.vmap(self.scoring_function, in_axes=(None, 0))(genotypes, jnp.asarray(subkeys))
...  #  adding to container and so on...
```

btjanaka replied:
I think it makes sense to have a scoring_fn that is the same as in MAPElites, but one issue I see is that the brax scoring function takes in a set of `init_states`. Using the vmap approach, it seems the `init_states` would be repeated across all the evals, thus resulting in 0 spread (IIRC the initial state is the only non-deterministic component in these environments). 

One potential idea (albeit a backwards-incompatible one) is that perhaps `scoring_function_brax_envs` should actually be taking in a series of random keys, one for each controller it is evaluating. That way, if you want to create different environments, you can pass in a ton of random keys, and if you don't want stochasticity, you pass in the same random key repeated several times.

limbryan replied:
Not off the top of my head (the tutorials need some updating in general but haven't had the time to get around it) but ```env.reset``` or ```partial(env.reset)``` should work as the ```play_reset_fn```

btjanaka replied:
@Lookatator @limbryan I finished implementing this change! I added a `multi_eval_scoring_function` in `MELS.__init__`. The scoring function now uses `reset_based_scoring_function_brax_envs`

Lookatator replied:
Great! Thank you @btjanaka!

Lookatator commented on qdax/core/mels.py:
```diff
@@ -0,0 +1,120 @@
+                scoring_function, in_axes=(None, 0)
+            )(genotypes, subkeys[1:])
+
+            # The vmap will output arrays with shape (num_evals, batch_size, ...)
+            # but we need to have (batch_size, num_evals, ...) for MELSRepertoire.
+            fitnesses = jax.tree_map(lambda x: jnp.swapaxes(x, 0, 1), fitnesses)
+            descriptors = jax.tree_map(lambda x: jnp.swapaxes(x, 0, 1), descriptors)
+            extra_scores = jax.tree_map(lambda x: jnp.swapaxes(x, 0, 1), extra_scores)
```
[Optional Tip] To avoid having to do this, I think you can simply add the argument `out_axes=1` to the `vmap` above.

Egiob commented on tests/core_test/containers_test/mels_repertoire_test.py:
```diff
@@ -0,0 +1,230 @@
+from qdax.types import ExtraScores
+
+
+def test_mels_repertoire() -> None:
```
Minor suggestion: we could use a short sentence to define the scope of the test

btjanaka replied:
Updated the test names and added docstrings

Egiob commented on tests/core_test/containers_test/mels_repertoire_test.py:
```diff
@@ -0,0 +1,230 @@
+        pytest.assume(jnp.allclose(repertoire.spreads, expected_spreads_2, atol=1e-6))
+
+
+def test_single_eval() -> None:
```
Minor suggestion: we could use a short sentence to define the scope of the test

btjanaka replied:
Updated the test names and added docstrings

Egiob commented on qdax/core/containers/mels_repertoire.py:
```diff
@@ -0,0 +1,311 @@
+from qdax.types import Centroid, Descriptor, ExtraScores, Fitness, Genotype, Spread
+
+
+def _dispersion(descriptors: jnp.ndarray) -> jnp.ndarray:
```
Minor suggestion: perhaps these functions belong elsewhere in a utils/ file ? I feel like they can be used by other methods as scores or metrics

btjanaka replied:
I think if it becomes something used in multiple places, perhaps it can then be refactored? Since it is a private function, it seems it can be moved around pretty easily.

Egiob commented on qdax/core/mels.py:
```diff
@@ -0,0 +1,120 @@
+        metrics_function: Callable[[MELSRepertoire], Metrics],
+        num_evals: int,
+    ) -> None:
+        def multi_eval_scoring_function(
```
Minor suggestion: put this function on top of the file with a `make_multi_eval_scoring_function` to reduce nested indents.

btjanaka replied:
Agreed

Egiob reviewed (approved):
Hey @btjanaka, thanks for the very-clean PR, amazing work! I've left a couple of small suggestions, but all good for me :+1:

btjanaka commented:
Thank you @Egiob! I've resolved all your comments except for the one about moving `_dispersion`

manon-but-yes commented:
Hi @btjanaka thank you for this PR! Apologies for the delay I just got back to work.

manon-but-yes commented on qdax/core/mels.py:
```diff
@@ -0,0 +1,129 @@
+)
+
+
+def multi_eval_scoring_function(
```
There is a file containing a similar function in qdax/utils/sampling.py, that is uses by ME-ES, maybe we want to uniform this? 
multi_eval_scoring_function() seems to be a subcase of sampling() that returns all the evaluations, while sampling returns the final average, so sampling() could just wrap multi_eval_scoring_function().

btjanaka replied:
Hi @manon-but-yes, thank you for the feedback! I have created a `multi_sample_scoring_function` in `sampling.py` and removed `multi_eval_scoring_function` from `mels.py`. I also changed `num_evals` to `num_samples` everywhere I could find it. Let me know what you think!

manon-but-yes replied:
It's perfect!
